.github/workflows/wheels.yml

```
@@ -190,17 +190,26 @@ jobs:
 
     - run: make sdist
 
-    - uses: actions/upload-artifact@v3
+    - uses: actions/upload-artifact@v4
       with:
-        name: dist
+        name: sdist
         path: dist/*.tar.gz
 
-  success:
-    permissions:
-      contents: none
+  pypi-publish:
+    if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
     needs: [build, windows, sdist]
     runs-on: ubuntu-latest
-    name: Wheels Successful
+    name: Upload release to PyPI
+    environment:
+      name: release-pypi
+      url: https://pypi.org/p/Pillow
+    permissions:
+      id-token: write
     steps:
-      - name: Success
-        run: echo Wheels Successful
+      - uses: actions/download-artifact@v4
+      - name: Merge artifacts
+        run: |
+          mkdir dist
+          mv */* dist
+      - name: Publish to PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
```

I'm expecting we'll need a few cycles to get the actual upload working, so let's add some temporary commits to publish to TestPyPI, and then rebase them out later:

```suggestion
      - name: Publish to PyPI
        uses: pypa/gh-action-pypi-publish@release/v1
        with:
          repository-url: https://test.pypi.org/legacy/
```
Plus we could comment out/replace the `if` to temporarily trigger.

Do we want to use a staging environment before publishing? If so, please let a repo admin know what we need there (I don't know, other than to call it `release-pypi`).

We can always remove it from this PR and add it later if we want.

---

FYI my PyPUG guide was updated with examples of Sigstore signatures a few months ago: https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/.

And also, I suggest using two separate jobs for PyPI and TestPyPI since they are different environments semantically and would benefit from having different GitHub Environment names and URLs, being displayed distinctly in the GH UI + the protections can be different (I normally have PyPI with required reviewers and TestPyPI w/o it)

---

Fair